llvm/lib/Target/SPIRV/SPIRVBuiltins.cpp

```diff
@@ -1182,16 +1205,37 @@ static bool generateExtInst(const SPIRV::IncomingCall *Call,
                  : SPIRV::OpenCLExtInst::fmax;
   }
 
+  Register ReturnTypeId = GR->getSPIRVTypeID(Call->ReturnType);
+  SmallVector<Register> Arguments;
+  unsigned ResultElementCount =
+      GR->getScalarOrVectorComponentCount(ReturnTypeId);
+  bool MayNeedPromotionToVec =
+      builtinMayNeedPromotionToVec(Number) && ResultElementCount > 1;
+  for (Register Argument : Call->Arguments) {
+    Register PromotedArg = Argument;
+    SPIRVType *ArgumentType = GR->getSPIRVTypeForVReg(Argument);
+    if (MayNeedPromotionToVec && ArgumentType != Call->ReturnType) {
+      PromotedArg = createVirtualRegister(Call->ReturnType, GR, MIRBuilder);
+      auto VecBroadcast = MIRBuilder.buildInstr(SPIRV::OpCompositeConstruct)
+                              .addDef(PromotedArg)
+                              .addUse(ReturnTypeId);
+      for (unsigned I = 0; I != ResultElementCount; ++I)
+        VecBroadcast.addUse(Argument);
+    }
+    Arguments.push_back(PromotedArg);
+  }
+
```

Comment:
This code is a little hard to read. Use `MayNeedPromotionToVec` in the middle of the loop make is harder to determine what difference it makes. Could this be moved to a function `getCallArguements(...)`? Then in the function if you can have a different path:

```
if (!MayNeedPromotionToVec)
  return SmallVector(Call->Arguments.begin(), Call->Arguments.end());

// do loop
```

This makes if obvious that if you are not promoting, then you are simply using the existing arguments.